### PR TITLE
Video widget

### LINKED ISSUE #1134: Improve Video Widget
Also includes #639

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
In reference to resolve #1134 & #639

### CODE REVIEW

File: collect_app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -511,4 +511,5 @@
     <string name="server_settings_title">Server</string>
     <string name="type">Type</string>
     <string name="form_entry">Uncheck to hide from Form Entry</string>
+    <string name="open">Open in external app</string>
```
Comment (lognaturel) [inline collect_app/src/main/res/values/strings.xml]: Calling this `open_external_video` or something like that will help with readability.

**Comment by lognaturel** [comment]
I like how you addressed the full screen option by adding "open in external app!" That gives users the opportunity to choose the existing, familiar experience if they need or to quickly verify that they have the correct video or preview it inline.

When I take a landscape video and view the form in portrait, it works perfectly and just like in your wireframe. With other orientation combinations, things get more complicated.

- portrait video with portrait phone: after recording and accepting the video or swiping to another question and back to a video question, I am vertically scrolled down where all I see is the video and the play buttons. I would expect it to be scrolled to the top instead. I'm guessing you did that because when it's scrolled to the top of the question, there's no way to play the video. In the case of a portrait video, would it be possible to scale it based on the available vertical space rather than scaling it according to the horizontal space available? Do you always know the video orientation?

- landscape phone: no matter the video orientation, there isn't much space below the buttons for the video to go. Maybe the video scale could always be optimized for portrait mode, even when it's displayed in landscape? That way the video would still be partly off the screen but at least more of it could be seen in one screenful. I don't find it very helpful to only see a tiny fraction of the video!

Other thoughts I had:
- the font size for the play buttons doesn't seem to match any other sizes in the app. I think one thing that will be important for a consistent look and feel will be using a limited set of font sizes. How about setting it explicitly with a new style so that you can more easily reuse the same one and change it later?
- how about making it so that tapping anywhere in the video surface plays or pauses?
- "Play Video" should probably be "Play video" now.

If you haven't already, you may find it helpful to take a look at the comments on #640 really quickly since many are related.

@nribeka @yanokwa You've thought about this before so it'd be great to get your impressions!

File: collect_app/src/main/res/values/strings.xml (hunk 1 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/res/values/strings.xml]: Done

**Comment by shobhitagarwal1612** [comment]
> how about making it so that tapping anywhere in the video surface plays or pauses?

Then should the `mediaController` be removed?

**Comment by lognaturel** [comment]
Can tapping on the video both toggle play/pause and show/hide the media controller?

**Comment by shobhitagarwal1612** [comment]
> Can tapping on the video both toggle play/pause and show/hide the media controller?

@lognaturel Done

**Comment by lognaturel** [comment]
Nice! Regarding scaling, here are some ideas we had discussed in #collect-code. I'm sure there are other possibilities. The goal is to minimize the cases in which a user would end up seeing only a piece of the video (it wouldn't fit on the available screen real estate).

1. always optimize the scale for portrait and keep the same size for landscape. 
So in my little doodle, in landscape, the 1x1 video would fill the height and be centered on the width. It would keep the exact same size as in portrait.

2. only show the preview in portrait and just use the same button as before in landscape.

3. scale down to some fixed, small size. I think this is dangerous because there's such a wide range of devices.

4. always scale the wide edge of the video to match the narrow edge of the device (I think this is the same as option 1) or some other relative scaling like that.

**Comment by shobhitagarwal1612** [comment]
@lognaturel I think option 1 is a better one. I've pushed the fix. Please review

**Comment by lognaturel** [comment]
I'm having trouble evaluating this because it's not clear to me exactly what should happen. I think it would be helpful to have a specification for the different device and video orientation combinations. If there's a device and Android version combination that you tried and that worked as intended, that would be helpful to know as well.

To give a sense of what I'm seeing, I would expect the widget to start scrolled at the top with buttons visible. Instead, it looks like it's scrolling down sometimes. Portrait video doesn't work at all on the Galaxy Tab A or the Moto G4 currently when I believe it did before. I'm now sometimes seeing a gray square background behind landscape videos but it doesn't look consistent.

Here is one possibility:
video
All sides with a little `o` around them are the same length and all sides with `"` are the same length. That would let the video occasionally go off the screen but it would have some consistency.

**Comment by shobhitagarwal1612** [comment]
@lognaturel I have made the modifications as you suggested. Please try it out and give reviews

**Comment by lognaturel** [comment]
Can you please write out a specification of what you think should happen? I'm still not totally clear on whether we are expecting the same thing. I was suggesting that the long edge of the video always match the short edge of the screen. That's what my (I'll admit not so nice!) sketch was trying to demonstrate. With the current implementation, it looks like with landscape video on a landscape phone and portrait video on a portrait phone the short edge of the video matches the short edge of the screen. I'm not necessarily against that but I want to make sure we are on the same page.

On all the devices and emulators I've tried, the screen scrolls such that the video always starts at the top of the screen and the question and buttons scroll off the top. I think you mentioned that this was not behavior you had seen. Do you have an API 22 emulator you could try it on?

I think that this scrolling behavior is confusing because it's different from all other question types. I would prefer that the question start scrolled to the very top as every other question does. That said, I could very well be wrong and what we could do is merge it in to `ui-refresh` and ask some actual enumerators.

**Comment by shobhitagarwal1612** [comment]
@lognaturel The video view by default requests focus (see [here](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/widget/VideoView.java#146)) because of which the screen gets scrolled down on orientation changes. I have overriden this behaviour by scrolling the screen up. I think you are right that this might confuse the enumerators. So what we can do is by default use the original behaviour of not focusing the video view. Then we can create a polling or something like that in Discourse where we can get feedback from the users regarding this behaviour. How does this sound to you?

**Comment by lognaturel** [comment]
Thanks, @shobhitagarwal1612, this works well on orientation change but it's still scrolled down when swiping back and forth into a video question.

**Comment by lognaturel** [comment]
I propose that we merge this as it is to `ui-refresh` as a proof of concept that we can get some real user feedback on. While putting audio inline struck me as a clear improvement, I don't find this nearly as clear. Before we invest more time trying to figure out the focus/scrolling issue, I think we should check with users, especially on smaller devices, to see if it's even an overall positive change.

@shobhitagarwal1612 @yanokwa how does that sound?

**Comment by shobhitagarwal1612** [comment]
Sounds good to me

**Pull request merged**